vmware/docker-volume-vsphere
# Handle default_datastore for vmgroup

This PR includes the most fix needed for #1143. For the detailed design, please refer to that issue.
**With this PR:**
1. _DEFAULT tenant will be created with two privileges (```__ALL_DS``` and ```__VM_DS```) after install.
2. "--default-datastore" is a required field for "vmgroup create" command
3. user can modify "default_datastore" through "vmgroup update" command
4. when "default_datastore" is set or modified, our code will try to create a privilege to the "default_datastore" with full access privilege if it does not exist
5. user cannot set "default_datastore" to ```__ALL_DS```, an error will be generated.
6. user cannot remove a privilege if the datastore in that privilege is the "default_datastore". An error will be generated.
7. change the vmdkops_test.py and vmdkops_admin_test.py accordingly

**Not covered in this PR,but needed to fix issue #1143 completely:**
1. upgrade related code: Without this fix, _DEFAULT tenant is created with ```default_privilege``` after install, we don't have the ```default_privilege``` any more, need to replace with two privileges ```__VM_DS``` and ```ALL_DS```
2. upgrade related code:  for any vmgroup, if the "default_datastore" is set, need to create a full access privilege to that datastore if it does not exist; and need to set the "default_datastore" to ```__VM_DS``` if it does not set since with the fix in this PR,  any "vmgroup" need set "default_datastore" field
3. The upgrade code is dependent on Shahzeb's PR, I plan to work on it after his PR is merged
4. Change the AdminCLI document accordingly


**Test Summary**

The following automated test has been modified or added for the new behavior of default_datastore handling introduced in this PR.

**vmdkops_test**

- test for default vmgroup
 1. DEFAULT vmgroup, privilege to datastore "__ALL_DS", and privilege to datastore "__VM_DS""
     are present post install, issue vmdk_ops from VM which is not owned by any vmgroup vmdk_ops 
     should succeed, and the volume create with short name will be created in the vm_datastore 
     where VM lives
  2. change the default_datastore for DEFAULT vmgroup to a real datastore,  volume create with short name should be created in the default datastore instead of VM datastore
  3. Remove privilege to "__ALL_DS" and let only privilege to  "__VM_DS" present, "create volume" with short name should succeed; while try to create a volume with full name like volume@datastore, and datastore is not the vm_datastore, the creation should fail
  4.  Remove privilege to "__VM_DS" and now no privilege available for _DEFAULT vmgroup, "create volume" should fail
  5. REMOVE DEFAULT vmgroup, "create volume" should fail

- test for named vmgroup
1. named vmgroup "tenant1" is created with default_datastore set to "VM_DS", test volume
create with short name and full name.  With short name, volume is created in the default_datastore which is the vm_datastore. Also test volume remove/attach/detach/ls.
 2. change the default_datastore to from "VM_DS" to a real datastore, and set the "volume_maxsize" an "usage_quota" to test volume create with different sizes to make sure
volume create can only succeed with the "volume_maxsize" and "usage_quota" limit.
 3. rename the "tenant11" to "new_tenant1", after that, test basic volume create/attach/detach/remove/ls

- test for vmdkops on two different vmgroups
1. Two vmgroups are created with "default_datastore" set to "__VM_DS"
2. For each vmgroup, update the "default_datastore" to a real datastore (self.datastore), and try to update the privilege to this datastore, test volume create with different sizes . Run volume ls to make sure each vmgroup can only see the volumes belong to that vmgroup

- test for default_datastore
1. Named vmgroup "tenant1" is created with "default_datastore" set to "VM_DS"
2. Test volume create with short name succeeds and the volume is created in the vm_datastore
3. Modify the privilege "VM_DS" to disallow create, then volume create with short name failed
4. set the "default_datastore" to a real datastore, make sure volume create succeeds and volume is created on the "default_datastore"
 5. modify the "default_datastore" to a different real datastore, make sure volume create succeeds and volume is created on the new "default_datastore"

**AminCLI test for "vmgroup" management**

- vmgroup create
1. “vmgroup create” without  “—default-datastore” failed
2. “vmgroup create” with –default-datastore=datastore1 succeed
3. “vmgroup create” with –default-datastore=__VM_DS succeed
4. “vmgroup create” with –default-datastore=__ALL_DS failed
5. “vmgroup create” with –default-datastore=invalid_datastore_name failed
 
 

- vmgroup update
1. “vmgroup update” with --default-datastore=valid_datastore_name succeeds
2. “vmgroup update” with –default-datastore=__ALL_DS failed
 

- vmgroup access add
1. add  “usage_quota” for privilege to “__ALL_DS” and “__VM_DS” is not allowed
 

- vmgroup access set
1. change  “usage_quota” for privilege to “__ALL_DS” and “__VM_DS” is not allowed
 
 

- vmgroup access rm
1. Remove privilege to “default_datastore” failed
